temporal-sdk/src/main/java/io/temporal/internal/worker/ShutdownManager.java

```diff
@@ -0,0 +1,161 @@
+/*
+ *  Copyright (C) 2020 Temporal Technologies, Inc. All Rights Reserved.
+ *
+ *  Copyright 2012-2016 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ *
+ *  Modifications copyright (C) 2017 Uber Technologies, Inc.
+ *
+ *  Licensed under the Apache License, Version 2.0 (the "License"). You may not
+ *  use this file except in compliance with the License. A copy of the License is
+ *  located at
+ *
+ *  http://aws.amazon.com/apache2.0
+ *
+ *  or in the "license" file accompanying this file. This file is distributed on
+ *  an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ *  express or implied. See the License for the specific language governing
+ *  permissions and limitations under the License.
+ */
+
+package io.temporal.internal.worker;
+
+import java.io.Closeable;
+import java.time.Duration;
+import java.util.concurrent.*;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class ShutdownManager implements Closeable {
+  private static final Logger log = LoggerFactory.getLogger(ShutdownManager.class);
+
+  private final ScheduledExecutorService scheduledExecutorService =
+      Executors.newSingleThreadScheduledExecutor(
+          new ExecutorThreadFactory(
+              WorkerThreadsNameHelper.SHUTDOWN_MANAGER_THREAD_NAME_PREFIX, null));
+
+  private static final int CHECK_PERIOD_MS = 500;
+
+  public CompletableFuture<Void> shutdownExecutorNow(
+      ExecutorService executorToShutdown, String executorName, Duration timeout) {
+    return shutdownExecutor(executorToShutdown, executorName, timeout, true);
+  }
+
+  public CompletableFuture<Void> shutdownExecutor(
+      ExecutorService executorToShutdown, String executorName, Duration timeout) {
+    return shutdownExecutor(executorToShutdown, executorName, timeout, false);
+  }
+
+  private CompletableFuture<Void> shutdownExecutor(
+      ExecutorService executorToShutdown,
+      String executorName,
+      Duration timeout,
+      boolean shutdownNow) {
+    if (shutdownNow) {
+      executorToShutdown.shutdownNow();
+    } else {
+      executorToShutdown.shutdown();
+    }
+
+    int attempts = (int) Math.ceil((double) timeout.toMillis() / CHECK_PERIOD_MS);
+
+    CompletableFuture<Void> future = new CompletableFuture<>();
+    scheduledExecutorService.scheduleWithFixedDelay(
```

Comment:
It will work and it does look better and more elegant.
But I have to disagree here that it's easier for a reader. I really want the task implementation itself to define the behavior fully, including when it should be canceled. Otherwise, to understand the logic of the task, you need to look into two places and make some mental gymnastics to understand when the cancel happens and that it actually happens at all.
Let me rewrite it a bit differently...